pkg/policy/mapstate.go

```diff
@@ -231,21 +315,21 @@ func (keys MapState) DenyPreferredInsert(newKey Key, newEntry MapStateEntry) {
 		return
 	}
 
-	keys.RedirectPreferredInsert(newKey, newEntry)
+	keys.RedirectPreferredInsert(newKey, newEntry, adds, deletes)
 }
 
 // RedirectPreferredInsert inserts a new entry giving priority to L7-redirects by
 // not overwriting a L7-redirect entry with a non-redirect entry.
-func (keys MapState) RedirectPreferredInsert(key Key, entry MapStateEntry) {
-	if !entry.IsRedirectEntry() {
-		if _, ok := keys[key]; ok {
-			// Key already exist, keep the existing entry so that
-			// a redirect entry is never overwritten by a non-redirect
-			// entry
-			return
-		}
+func (keys MapState) RedirectPreferredInsert(key Key, entry MapStateEntry, adds, deletes MapState) {
+	// Do not overwrite the entry, but only merge selectors if the old entry is a deny or redirect.
+	// This prevents an existing deny or redirect being overridden by a non-deny or a non-redirect.
+	if oldEntry, exists := keys[key]; exists && (oldEntry.IsRedirectEntry() || oldEntry.IsDeny) {
+		oldEntry.MergeSelectors(&entry)
```

Comment:
A comment would help here to explain why merging is required. It's not entirely obvious. Does the entry cover both selectors of redirects and non-redirects. What happens when the selector of the redirect issues a delete? Who transforms the entry to a non-redirect? Is that always done with a full policy recalculation?

Reply:
Currently the transformation back to non-redirect is done only on full policy recalculation (i.e., on policy or configuration changes), and proxy redirection will stick around until then and the packets not needing the proxy redirection will keep on being redirected like they were in the presence of the redirect entry. I'll comment on this, but a a bit more involving design is needed to get rid of this stickiness. Maybe like keeping multiple entries around in the MapState in priority order so that the one with the next highest precedence can be instantiated when an entry is removed?

Reply:
OK. We don't have to handle this as part of this PR at all.